# Nicaragua country script

This script is still a work in progress and is somewhat outdated

In [1]:
import os, sys, glob, json
from itertools import product, compress, chain
from functools import reduce
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)
import requests
import numpy as np
import pandas as pd
import geopandas as gpd
from tools import save_hdf, save_npz, load_npz, PrintAdminUnits, PlotAdminShapes
from tools import FDW_PD_Sweeper, FDW_PD_AvalTable, FDW_PD_Compiling, FDW_PD_GrainTypeAgg, FDW_PD_ValidateFnidName
from tools import FDW_PD_CreateAdminLink, FDW_PD_RatioAdminLink, FDW_PD_ConnectAdminLink
from tools import FDW_PD_CaliSeasonYear
from tools_graphic import PlotBarProduction, PlotLinePAY, PlotHeatCropSystem, PlotHeatSeasonData
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
warnings.simplefilter(action='ignore', category=pd.errors.PerformanceWarning)
pd.options.mode.chained_assignment = None


In [2]:
# CPCV2 grain code ------------------------------ #
grain_code = pd.read_hdf('../data/crop/grain_cpcv2_code.hdf')
product_category = grain_code[['product', 'product_category']].set_index('product').to_dict()['product_category']
# ----------------------------------------------- #

# Load FEWS NET administrative boundaries ------- #
epsg = 'epsg:32617' # 
fn_shapes = sorted(glob.glob('../data/shapefile/fewsnet/NI_Admin?_????.shp'))
shape_all = []
for fn in fn_shapes:
    name = fn[-18:-4]
    fn = fn.replace('\\', '/')
    exec('%s = gpd.read_file("%s").to_crs("%s")' % (name, fn, epsg))
    exec('%s["area"] = %s["geometry"].area/10**6' % (name, name))
    exec('shape_all.append(%s)' % (name))
shape_all = pd.concat(shape_all, axis=0).reset_index(drop=True)
PrintAdminUnits(shape_all)
# ----------------------------------------------- #

- FEWS NET admin shapefiles ------------------- #
| year	 | Admin1   | # units   | Admin2   | # units   | Admin3   | # units   |
| 1998	 | NI1998A1 | 17	| nan	| 0	| nan	| 0	|
| 2005	 | NI2005A1 | 17	| NI2005A2	| 152	| nan	| 0	|
----------------------------------------------- #


In [3]:
# FDW API host address -------------------------- #
host = 'https://fdw.fews.net'
auth = tuple(json.loads(open('token.json', "r").read()))
parameters = {
    'format': 'json',
    'country': 'Nicaragua',
    'product': 'R011',
    'survey_type': 'crop:best'
}
endpoint = '/api/cropproductionindicatorvalue/'
response = requests.get(host + endpoint, auth=auth, params=parameters, proxies={})
response.raise_for_status()
df = pd.DataFrame.from_records(response.json())
df_origin = df.copy()
# ----------------------------------------------- #


In [4]:
# Manual Pre-processing before Sweeping --------- #
# 1. Default setting
# a) None-type crop production system
df.loc[df['crop_production_system'].isna(), 'crop_production_system'] = 'none'
df.loc[df['crop_production_system'] == '', 'crop_production_system'] = 'none'
# b) None-type population group
df.loc[df['population_group'].isna(), 'population_group'] = 'none'
df.loc[df['population_group'] == '', 'population_group'] = 'none'

# 2. Manual setting
# a) Add admin names to reporting units
admin_level = 'admin_1'
reporting_unit = df.loc[df['fnid'].map(lambda x: x[6] != 'A'), 'fnid'].unique()
df.loc[df.fnid.isin(reporting_unit),admin_level] = df.loc[df.fnid.isin(reporting_unit), 'locality_name'].map(lambda x: x.split(',')[0])

# FDW Production Data Inspection ---------------- #
df, df_raw = FDW_PD_Sweeper(df)
table_dict = FDW_PD_AvalTable(df, shape_all)
# ----------------------------------------------- #

# FEWS NET Shapefile comparison ----------------- #
shape_used = pd.concat([NI_Admin1_1998, NI_Admin1_2005], axis=0)
PlotAdminShapes(shape_used, label=True)
# ----------------------------------------------- #

- Remove missing records ---------------------- #
Orignial data points: 2,052
Removed 186 "Missing Value" points
684/684 "Area Harvested" points are retained.
591/684 "Quantity Produced" points are retained.
591/684 "Yield" points are retained.
Current data points: 1,866

- Minor changes are applied ------------------- #

- Basic information --------------------------- #
Data period: 2000 - 2012
7 grain types are found: Maize (Corn), Maize Grain (White), Rice (Paddy), Sorghum, Sorghum (Red), Sorghum (White), Sorghum grain (Millón)
5 seasons are found: Annual (12-01), Postrera (09-01), Secano (05-01), Primera (05-01), Riego (06-01)
2 crop production system are found: none, irrigated
Data sources include:
[1] Ministerio Agropecuario y Forestal Magfor, Nicaragua (replaced by 144) --- Crop Production Survey, Nicaragua
Administrative-1 fnids: 17
Administrative-2 fnids: 0
0 reporting units are found: 

- Total production over time ------------------ #
season_name         Annual
             

## Nicaragua
Nicaragua crop seasonal calendar

<img src="https://fews.net/sites/default/files/styles/large_width_880/public/2023-03/seasonal-calendar-nicaragua%20%281%29.png?itok=GThlj6xQ" width=900></br>

<img src="https://ipad.fas.usda.gov/rssiws/al/crop_calendar/images/ca_nu_calendar.png" width=900>

- comparison of boundaries

<img src="../figures/NI_admin_shapes.png" width=900>

- FDW data consists of `NI1998A1`.

- FEWS NET admin shapefiles ------------------- #

| year	 | Admin1   | # units   | Admin2   | # units   | Admin3   | # units   |
| :---: | :----:  | :----:   | :----:   | :---:  |:---:  |:---:  |
| 1998	 | NI1998A1 | 17	| nan	| 0	| nan	| 0	|
| 2005	 | NI2005A1 | 17	| NI2005A2	| 152	| nan	| 0	|

- **`NI1998A1`** is used to represent the current admin-level 1 crop data.
- Nicaragua has five crop seasons: `Annual`, `Postrera`, `Secano`, `Primera`, `Riego`, although only `Annual` has more than one year of data. 
- Nicaragua has two crop production systems: `none`, `irrigated`.

In [5]:
# Define the latest shapefile ------------------- #
latest_level = 1
shape_latest = NI_Admin1_1998.copy().to_crs('epsg:4326')
# ----------------------------------------------- #

# Validation of FNIDs and Names ----------------- #
df = FDW_PD_ValidateFnidName(df, shape_used, shape_latest)
# ----------------------------------------------- #

# FDW Production Data Compiling ----------------- #
area, prod = FDW_PD_Compiling(df, shape_used)
area_all, prod_all = area.copy(), prod.copy()
mdx_pss = area.columns.droplevel([0,1]).unique()
# ----------------------------------------------- #

In [6]:
# Link admin boundaries ------------------------- #
link, over = FDW_PD_CreateAdminLink(NI_Admin1_1998, NI_Admin1_1998, 'ADMIN1', 'ADMIN1', prod, epsg)
# Crop specific ratios
link_ratio = FDW_PD_RatioAdminLink(link, prod, over, mdx_pss)
# Add current unit to link_ratio
for fnid_new in link_ratio.keys():
    link_ratio[fnid_new][fnid_new] = 1.0
    link_ratio[fnid_new] = link_ratio[fnid_new].sort_index(axis=1, ascending=False)
# Connect data with AdminLink
area_new, prod_new = FDW_PD_ConnectAdminLink(link_ratio, area, prod, validation=False)
# ----------------------------------------------- #

# Manual correction ----------------------------- #
crop_new = prod_new/area_new

# Complete long format DataFrame ---------------- #
df_area = area_new.T.stack().rename('value').reset_index()
df_area['indicator'] = 'area'
df_prod = prod_new.T.stack().rename('value').reset_index()
df_prod['indicator'] = 'production'
df_yield = crop_new.T.stack().rename('value').reset_index()
df_yield['indicator'] = 'yield'
stack = pd.concat([df_area, df_prod, df_yield], axis=0)
# Add "planting year"
cols = ['season_name','product','crop_production_system','planting_month','harvest_year','harvest_month','planting_year']
season_table = df[cols].drop_duplicates()
stack = stack.merge(season_table, on=cols[:-1])
# Add country and admin names
stack = stack.merge(df[['fnid','country','country_code','admin_1','admin_2']].drop_duplicates(), on='fnid', how='inner')
names = [
    'fnid','country','country_code','admin_1','admin_2','name',
    'product','season_name','planting_year','planting_month','harvest_year','harvest_month',
    'crop_production_system','indicator','value'
]
stack_gscd = stack[names]
stack_gscd['gscd_code'] = 'calibrated'
# ----------------------------------------------- #

# Reported FDW data ----------------------------- #
stack_fdw = df[names]
stack_fdw['indicator'] = stack_fdw['indicator'].replace({'Area Harvested':'area','Quantity Produced':'production','Yield':'yield'})
stack_fdw['gscd_code'] = 'reported'
# ----------------------------------------------- #

# Final Processing ------------------------------ #
stack = pd.concat([stack_fdw,stack_gscd], axis=0).reset_index(drop=True)
# No concerns found for grain types
stack['product'] = stack['product'].replace(product_category)
# Calibration of planting and Harvest year and season
cs = {
    'Annual': {'planting_month':{'12-01':'12-01'},'harvest_month':{'12-01':'12-01'}},
    'Secano': {'planting_month':{'05-01':'05-01'},'harvest_month':{'12-01':'12-01'}},
    'Postrera': {'planting_month':{'09-01':'09-01'},'harvest_month':{'12-01':'12-01'}},
    'Primera': {'planting_month':{'05-01':'05-01'},'harvest_month':{'08-01':'08-01'}},
    'Riego': {'planting_month':{'06-01':'06-01'},'harvest_month':{'02-01':'02-01'}},
}
cy = {}
stack, df, link_ratio = FDW_PD_CaliSeasonYear(stack, df, link_ratio, cs, cy)
stack.loc[stack['admin_1'].isna(), 'admin_1'] = 'none'
stack.loc[stack['admin_2'].isna(), 'admin_2'] = 'none'
# ----------------------------------------------- #

#define a QC flag field
# 0 = no flag
# 1 = outlier
stack['QC_flag'] = 0

########################################################################

# Save data
save_hdf('../data/crop/adm_crop_production_NI.hdf', stack)
save_hdf('../data/crop/adm_crop_production_NI_raw.hdf', df)
save_npz('../data/crop/adm_crop_production_NI_ratio.npz', link_ratio)

TypeError: FDW_PD_CaliSeasonYear() takes from 2 to 3 positional arguments but 5 were given

In [ ]:
# Bar chart of national grain production
country_iso, country_name = 'NI', 'Nicaragua'
df = pd.read_hdf('../data/crop/adm_crop_production_%s.hdf' % country_iso)

df = df[df['gscd_code']=='calibrated']; df['year'] = df['harvest_year']
year = [df['year'].min(), df['year'].max()]
product_order = ['Sorghum','Maize','Rice']
for season_name in ['Annual', 'Secano', 'Postrera', 'Primera', 'Riego']:
    footnote = 'National grain production in %s - %s' % (country_name, season_name)
    fn_save = '../figures/%s_bar_natgrainprod_%s.png' % (country_iso, season_name)
    sub = df[df['season_name'] == season_name]
    if 'production' not in sub['indicator'].unique():
        continue
    fig = PlotBarProduction(sub, year, product_order, footnote, fn_save)
    fig.show()


In [ ]:
# Lineplot of Production-Area-Yield (PAY) time-series
country_iso, country_name = 'NI', 'Nicaragua'
df = pd.read_hdf('../data/crop/adm_crop_production_%s.hdf' % country_iso)

df = df[df['gscd_code']=='calibrated']; df['year'] = df['harvest_year']
year = [df['year'].min(), df['year'].max()]
product_season = [
    ['Maize','Primera'],
    ['Maize','Annual'],
    ['Rice','Secano'],
    ['Rice','Riego'],
    ['Rice','Annual'],
    ['Sorghum grain (Millón)','Postrera'],
    ['Sorghum (Red)','Postrera'],
    ['Sorghum (White)','Postrera'],
    ['Sorghum','Annual'],
]
for product_name, season_name in product_season:
    footnote = 'Production-Area-Yield (PAY) time-series of %s - %s - %s' % (country_iso, product_name, season_name)
    fn_save = '../figures/%s_line_pay_%s_%s.png' % (country_iso, product_name, season_name)
    sub = df[(df['product'] == product_name) & (df['season_name'] == season_name)]
    if len(sub)==0:
        continue
    fig = PlotLinePAY(sub, year, footnote, fn_save)
    fig.show()


In [ ]:
# Calibrated PAY time-series per FNID
from tools_graphic import PlotLineCropTS
country_iso, country_name = 'NI', 'Nicaragua'
df = pd.read_hdf('../data/crop/adm_crop_production_%s.hdf' % country_iso)
df = df[df['gscd_code']=='calibrated']; df['year'] = df['harvest_year']
link_ratio = load_npz('../data/crop/adm_crop_production_%s_ratio.npz' % country_iso)
year_all = np.arange(df['year'].min(), df['year'].max()+1)
prod_season = [
    ['Maize', 'Primera'],
    ['Maize', 'Annual'],
    ['Rice', 'Secano'],
    ['Rice', 'Riego'],
    ['Rice', 'Annual'],
    ['Sorghum grain (Millón)', 'Postrera'],
    ['Sorghum (Red)', 'Postrera'],
    ['Sorghum (White)', 'Postrera'],
    ['Sorghum', 'Annual'],
]
for product_name, season_name in prod_season:
    sub = df[
        (df['product'] == product_name) &
        (df['season_name'] == season_name)
    ]
    for fnid in sub['fnid'].unique():
        sub_fps = sub[sub['fnid'] == fnid]
        fn_save = '../figures/crop_calibrated/%s_%s_%s_%s.png' % (country_iso, product_name, season_name, fnid)
        fig = PlotLineCropTS(sub_fps, fnid, product_name, season_name, link_ratio, year_all, fn_save)

